src/VisualStudio/Core/Def/ProjectSystem/Legacy/SolutionEventsBatchScopeCreator.cs

```diff
@@ -4,245 +4,224 @@
 {
+    /// <summary>
+    /// A lock for mutating all objects in this object. This class isn't expected to have any "interesting" locking requirements, so this should just be acquired
+    /// in all methods.
+    /// </summary>
+    private readonly object _gate = new object();
     private readonly List<(ProjectSystemProject project, IVsHierarchy hierarchy, ProjectSystemProject.BatchScope batchScope)> _fullSolutionLoadScopes = [];
 
-    private readonly IThreadingContext _threadingContext = threadingContext;
-    private readonly IServiceProvider _serviceProvider = serviceProvider;
-
+    /// <summary>
+    /// The cookie for our subscription to the running document table. Null if we're not currently subscribed.
+    /// </summary>
     private uint? _runningDocumentTableEventsCookie;
     private bool _isSubscribedToSolutionEvents = false;
     private bool _solutionLoaded = false;
 
+    private readonly IServiceProvider _serviceProvider = serviceProvider;
+
     public void StartTrackingProject(ProjectSystemProject project, IVsHierarchy hierarchy)
     {
-        _threadingContext.ThrowIfNotOnUIThread();
-
-        EnsureSubscribedToSolutionEvents();
-
-        if (!_solutionLoaded)
+        lock (_gate)
         {
-            _fullSolutionLoadScopes.Add((project, hierarchy, project.CreateBatchScope()));
+            EnsureSubscribedToSolutionEvents();
+
+            if (!_solutionLoaded)
+            {
+                _fullSolutionLoadScopes.Add((project, hierarchy, project.CreateBatchScope()));
 
-            EnsureSubscribedToRunningDocumentTableEvents();
+                EnsureSubscribedToRunningDocumentTableEvents();
+            }
         }
     }
 
     public void StopTrackingProject(ProjectSystemProject project)
     {
-        _threadingContext.ThrowIfNotOnUIThread();
-
-        foreach (var scope in _fullSolutionLoadScopes)
+        lock (_gate)
         {
-            if (scope.project == project)
+            foreach (var scope in _fullSolutionLoadScopes)
             {
-                scope.batchScope.Dispose();
-                _fullSolutionLoadScopes.Remove(scope);
-                break;
+                if (scope.project == project)
+                {
+                    scope.batchScope.Dispose();
+                    _fullSolutionLoadScopes.Remove(scope);
+                    break;
+                }
             }
-        }
 
-        EnsureUnsubscribedFromRunningDocumentTableEventsIfNoLongerNeeded();
+            EnsureUnsubscribedFromRunningDocumentTableEventsIfNoLongerNeeded();
+        }
     }
 
-    private void StopTrackingAllProjects()
+    private Task StopTrackingAllProjectsAsync(CancellationToken cancellationToken)
     {
-        _threadingContext.ThrowIfNotOnUIThread();
+        ImmutableArray<Task> batchScopeTasks;
 
-        foreach (var (_, _, batchScope) in _fullSolutionLoadScopes)
+        lock (_gate)
```

Comment:
can review later.  but do you want to consider semaphoreslim here or soemthing cancellable since you're in a task-based method now?

Reply:
Got clarification from the API owners: the cancellation could be triggered when we're aborting a solution load. In this case, I think it's simplest for us to ensure we clear out any batches we created, but we don't need to wait on those completing. I don't think that's critical to do anything here in that case: any legacy projects will still be unloaded and will still ensure their batches get removed one-by-one, but I imagine that's a bit surprising if somebody had to debug this. So added some comments and changed where the CancellationToken is used for clarity.